jax/lax/lax_control_flow.py

```diff
@@ -977,4 +975,130 @@ def _root_batch(args, dims, **params):
 xla.initial_style_translations[root_p] = xla.lower_fun(_root_impl, initial_style=True)
-batching.primitive_batchers[root_p] = _root_batch
+batching.deftraced(root_p)
+
+
+_Solves = collections.namedtuple('_Solves', 'forward tangent cotangent')
+
+def linear_solve(matvec, b, forward_solve, tangent_solve=None,
+                 cotangent_solve=None, symmetric=False):
+  """Differentiably solve the linear map matvec(x)=b for x.
+
+  Required invariant:
+      x = solve(matvec, b)
+      error = matvec(x) - b
+      assert all(error == 0)
+  """
+  if tangent_solve is None:
+    tangent_solve = forward_solve
+  if cotangent_solve is None:
+    cotangent_solve = tangent_solve
+
+  b_flat, in_args_tree = tree_flatten((b,))
+  b_avals = tuple(_map(_abstractify, b_flat))
+  jaxpr, consts, out_tree = _initial_style_jaxpr(matvec, in_args_tree, b_avals)
+
+  in_tree, = treedef_children(in_args_tree)
+  if in_tree != out_tree:
+    raise TypeError(
+        _tree_error_template("matvec", "b").format(out_tree, in_tree)
+    )
+
+  solve = _Solves(
+      _flatten_high_level_func(
+          forward_solve, in_tree, _tree_error_template("forward_solve", "b")),
+      _flatten_high_level_func(
+          tangent_solve, in_tree, _tree_error_template("tangent_solve", "b")),
+      _flatten_high_level_func(
+          cotangent_solve, in_tree, _tree_error_template("cotangent_solve", "b")),
+  )
+
+  out_flat = linear_solve_p.bind(
+      *itertools.chain(consts, b_flat), num_consts=len(consts), jaxpr=jaxpr,
+      solve=solve, symmetric=symmetric)
+  return tree_unflatten(out_tree, out_flat)
+
+
+def _linear_solve_abstract_eval(*args, **kwargs):
+  return args[kwargs['num_consts']:]
+
+
+def _linear_solve_impl(*args, **kwargs):
+  num_consts, jaxpr, solve, _ = split_dict(
+      kwargs, ['num_consts', 'jaxpr', 'solve', 'symmetric'])
+  params, b = split_list(args, [num_consts])
+  matvec = partial(core.jaxpr_as_fun(jaxpr), *params)
+  return solve.forward(matvec, *b)
+
+
+def _tangent_linear_map(func, params, params_dot, *x):
+  """Compute the tangent of a linear map.
+
+  Assuming ``func(*params, *x)`` is linear in ``x`` and computes ``A @ x``,
+  this function computes ``∂A @ x``.
+  """
+  zeros = [ad_util.zero] * len(x)
+  _, out_tangent = ad.jvp(lu.wrap_init(func)).call_wrapped(
+      params + list(x), params_dot + zeros)
+  return out_tangent
+
+
+def _linear_solve_jvp(primals, tangents, num_consts, jaxpr, solve, symmetric):
+  # A x - b = 0
+  # ∂A x + A ∂x - ∂b = 0
+  # ∂x = A^{-1} (∂b - ∂A x)
+
+  x = linear_solve_p.bind(
+      *primals, num_consts=num_consts, jaxpr=jaxpr, solve=solve,
+      symmetric=symmetric)
+
+  params, _ = split_list(primals, [num_consts])
+  params_dot, b_dot = split_list(tangents, [num_consts])
+
+  matvec = partial(core.jaxpr_as_fun(jaxpr), *params)
+  matvec_dot = partial(
+      _tangent_linear_map, core.jaxpr_as_fun(jaxpr), params, params_dot
+  )
+
+  if all(tangent is ad_util.zero for tangent in params_dot):
```

Comment:
What happens if `params_dot` or `b_dot` are a mixture of zero and not-zero? Seems like it might blow up.

Reply:
This should be handled now. I'm not entirely happy with the test coverage, but struggled to figure out how to come up with a test case that could involve a mixture of zeros/non-zeros -- I think I would need a three argument primitive?